### ISSUE TITLE
Get modularity score after community detection in leiden/louvain

### ISSUE THREAD

**Issue by AlSzmigiel** [scverse/scanpy#2908]
### What kind of feature would you like to request?

Additional function parameters / changed functionality / changed defaults?

### Please describe your wishes

This was already discussed in https://github.com/scverse/scanpy/pull/819, is there any reason why it is not possible now?

**Comment by ivirshup** [comment]
I'm still dubious of the value, especially when we provide different ways of ways of optimizing the score.

What would you think of instead having `sc.metrics.modularity` where you match a clustering and a graph returning a modularity score?

It would basically wrap:

```python
(
    igraph.Graph.Weighted_Adjacency(adata.obsp["connectivities"])
    .modularity(adata.obs["louvain"].cat.codes)
)
```

But you could also generate modularity scores for other labelings.

**Comment by AlSzmigiel** [comment]
What about comparing communities between for example CPM and RBERVertexPartition at the same resolution, using modularity score? This way we are not comparing scores obtained by optimization functions, just simple "external" measure.

**Comment by ivirshup** [comment]
Yeah, that's why I like having it be manually computed rather than relying on the output of the clustering. This way we can make sure similar scores are used.

For API, I'm thinking:


```python
@singledispatch
def modularity(graph: sparse.spmatrix, labels: pd.Series | np.ndarray) -> float:
    ...

@modularity.register
def _modularity_adata(adata: AnnData, labels: pd.Series | str, obsp: str) -> float:
    return modularity(
        adata.obsp[obsp],
        adata.obs[labels] if isinstance(labels, str) else labels,
    )
```

**Comment by jpintar** [comment]
I enthusiastically support adding a `sc.metrics.modularity` function, but precisely because there are multiple ways of optimising the score to choose from, it would be useful to record the score actually achieved for a clustering by whichever method was selected to produce it.  One could then note in the documentation that this modularity value may differ from the one computed by the `igraph`-based `sc.metrics.modularity`, esp. if `igraph` wasn't used to generate the clustering.

As things stand, that value is already computed and is effectively being thrown away. I don't see what harm it would do to add `modularity = part.modularity` after [line 189 here](https://github.com/scverse/scanpy/blob/866535873a9b0cd645ccc285400c2d33941e459e/src/scanpy/tools/_leiden.py#L189) and then  to save it in `adata.uns["leiden"]` (I'm happy to write a little PR for that, if @ivirshup agrees).

**Comment by flying-sheep** [comment]
Oops, seems like the version I shipped in 1.12.0 has a bug, fixing in #4045
